src/components/twilio-voice-dialer/public/twilio-voice-dialer.js

```diff
@@ -1 +1,51 @@
-// TODO: Implement Web Component
+const render = (x) => `
+  <div class="container">
+    <p id="status">Status: pending</p>
+    <input type="text" placeholder="recipient" id="recipient" value=${x.recipient}>
+  </div>
+`;
+
+class TwilioVoiceDialer extends HTMLElement {
+  static get observedAttributes() {
+    return ['recipient'];
+  }
+
+  #device;
+  #token;
+
+  constructor() {
+    super();
+    this.attachShadow({ mode: 'open' });
+  }
+  get recipient() {
+    return this.getAttribute('recipient');
+  }
+
+  connectedCallback() {}
+
+  disconnectedCallback() {
+    this.#device.destroy();
+  }
+
+  attributeChangedCallback() {
+    this.shadowRoot.innerHTML = render(this);
```

Comment:
This seems like an anti-pattern. I think we should be constructing elements and appending them instead of using `innerHTML`.
```
const containerDiv = document.createElement('div');
// ... create and append the elements as necessary
this.shadowRoot.appendChild(containerDiv);
```

Reply:
The biggest drawback I can think of is, the browser will need to go through each node to render the whole thing. But with appending elements, we can update specific nodes. However, the advantage I see is, it's easy to understand the whole thing.

But as the whole project gets bigger, none of these approach is good. It's better to use a rendering engine at that point similar to how React, Angular, or Vue.

Since this is a pretty small project, I don't have a preference unless there's a security concern. You both can decide this one :)

Reply:
`webcomponents.org` uses this pattern
```
const header = document.createElement('header');
const shadowRoot = header.attachShadow({mode: 'open'});
shadowRoot.innerHTML = '<h1>Hello Shadow DOM</h1>'; // Could also use appendChild().
// header.shadowRoot === shadowRoot
// shadowRoot.host === header
```
https://www.webcomponents.org/introduction#creating-and-using-a-shadow-root

Reply:
Also mdn: https://developer.mozilla.org/en-US/docs/Web/API/ShadowRoot/innerHTML
```
let customElem = document.querySelector("my-shadow-dom-element");
let shadow = customElem.shadowRoot;

shadow.innerHTML = "<strong>This element should be more important!</strong>";
```

Reply:
I'm a little surprised that the official docs use `innerHTML` since using it is typically a code-smell. To be clear, the way that we're using it is perfectly safe as far as I can tell, it's just that when people do typically use it, they use it in an unsafe manner and therefore just not using it at all is safer. So in the end, I think it's fine then.